test/test.ts

```diff
@@ -0,0 +1,272 @@
+import {expect} from "chai"
+import {ConditionalOperator, MethodName, SqlJsonPathStatement} from "../src/json-path";
+import {compile} from "../src";
+
+interface SqlJsonPathTest {
+  statement: string
+  expected: SqlJsonPathStatement
+}
+
+describe('SQL Json Path', function () {
+  describe('parser test', function () {
+    const tests: SqlJsonPathTest[] = [
+      {
+        statement: "$.foo.bar",
+        expected: {
+          lhs: [
+            {
+              property: "foo"
+            },
+            {
+              property: "bar"
+            },
+          ]
+        }
+      },
+      {
+        statement: "$.foo.bar[*].location",
+        expected: {
+          lhs: [
+            {
+              property: "foo"
+            },
+            {
+              array: "bar",
+              element: "*"
+            },
+            {
+              property: "location"
+            },
+          ]
+        }
+      },
+      {
+        statement: "$.foo.bar[1].location",
+        expected: {
+          lhs: [
+            {
+              property: "foo"
+            },
+            {
+              array: "bar",
+              element: 1
+
+            },
+            {
+              property: "location"
+            },
+          ]
+        }
+      },
+      {
+        statement: "$.foo.bar[*].zar ? (@ > 333)",
+        expected: {
+          lhs: [
+            {
+              property: "foo"
+            },
+            {
+              array: "bar",
+              element: "*"
+            },
+            {
+
+              property: "zar",
+              filterChain: [{
+                query: {
+                  path: [],
+                  operator: ConditionalOperator.GT,
+                  compareTo: 333
+                }
+              }]
+            },
+          ]
+        }
+      },
+      {
+        statement: "$.foo.bar[*] ? (@.value != 22).zoo",
+        expected: {
+          lhs: [
+            {
+              property: "foo"
+            },
+            {
+              array: "bar",
+              element: "*",
+              filterChain: [{
+                query: {
+                  operator: ConditionalOperator.NE,
+                  compareTo: 22,
+                  path: [
+                    {
+                      property: "value"
+                    }
+                  ]
+                }
+              }]
+            },
+            {
+              property: "zoo"
+            }
+          ]
+        }
+      },
+      {
+        statement: "$.foo.bar[*] ? (@.zoo[1] < 13) ? (@.roo > 130).goo",
```

Comment:
can you really have multiple ? statements? The spec for it is in that Sql-2016 PDF, section `6.13 Filter expression`

Reply:
Yes, I copied these examples straight from: https://www.postgresql.org/docs/13/functions-json.html#FUNCTIONS-SQLJSON-PATH

You can chain and nest filter expressions.

Reply:
That's pretty nice! I am glad you noticed that.

Reply:
It's bugging me, though, because I don't think that is part of the SQL2016 spec. I looked at Oracle, MySql and SqlServer's jsonpath docs and they don't show that one can use more than one `?` in a filter statement. For instance, Oracle's docs: https://docs.oracle.com/en/database/oracle/oracle-database/21/adjsn/condition-JSON_EXISTS.html#GUID-D60A7E52-8819-4D33-AEDB-223AB7BDE60A

Reply:
I'm thinking of setting up some docker instances of the different databases and trying out various path statements to see what is supported. I know Postgres went beyond the spec (they support ** which isn't part of the spec, for instance).

Reply:
I found a free online BNF reference for SQL2016: https://jakewheat.github.io/sql-overview/sql-2016-foundation-grammar.html#_6_34_json_query

Reply:
Also highly relevant: https://jakewheat.github.io/sql-overview/sql-2016-foundation-grammar.html#_9_38_sqljson_path_language_lexical_elements